src/librustc_driver/args/mod.rs

```diff
@@ -0,0 +1,207 @@
+/// backslash. The file may itself contain additional @file options; any such options will be
+/// processed recursively.
+///
+/// See https://sourceware.org/binutils/docs/ld/Options.html#Options for the documentation of
+/// gnu-ld's @file parameters.
+pub struct ArgsStack {
+    base: env::ArgsOs,
+    stack: Vec<Box<dyn Iterator<Item = FileArg>>>,
+}
+
+impl ArgsStack {
+    pub fn new() -> Self {
+        ArgsStack { base: env::args_os(), stack: vec![] }
+    }
+
+    fn stack_next(&mut self) -> Option<FileArg> {
+        while let Some(top) = self.stack.last_mut() {
+            match top.next() {
+                Some(filearg) => return Some(filearg),
+                None => {
+                    let _ = self.stack.pop();
+                }
+            }
+        }
+
+        None
+    }
+}
+
+impl Iterator for ArgsStack {
+    type Item = OsString;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let arg = self.stack_next().or_else(|| self.base.next().map(FileArg::from));
+            match arg {
+                None => return None,
+                Some(FileArg::Arg(arg)) => return Some(arg),
+                Some(FileArg::Path(path)) => {
+                    let file = if self.stack.len() >= MAX_RECURSION {
+                        eprintln!(
+                            "Warning: Max recursion depth {} reached for file {}",
+                            MAX_RECURSION,
+                            path.display()
+                        );
+                        None
+                    } else {
+                        match fs::read(&path) {
+                            Ok(file) => Some(file),
+                            Err(err) => {
+                                eprintln!(
+                                    "Warning: Failed to open {}: {} (cwd: {})",
+                                    path.display(),
+                                    err,
+                                    env::current_dir().unwrap().display(),
+                                );
+                                None
+                            }
+                        }
+                    };
+                    if let Some(file) = file {
+                        self.stack.push(Box::new(file_args(file)))
+                    } else {
+                        // Failed to open - try to reconstruct original arg
+                        let arg = format!("@{}", path.display());
+                        return Some(OsString::from(arg));
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[cfg(any(unix, target_os = "redox"))]
+fn os_string_from_bytes(s: &[u8]) -> OsString {
+    use std::os::unix::ffi::OsStrExt;
+    std::ffi::OsStr::from_bytes(s).to_os_string()
+}
+#[cfg(not(any(unix, target_os = "redox")))]
+fn os_string_from_bytes(s: &[u8]) -> OsString {
+    OsString::from(&*String::from_utf8_lossy(s))
```

Comment:
I don't think that `from_utf8_lossy` is the right one to use here, I think we'll want to give a first-class error if the file isn't utf-8 on Windows.

Reply:
A more general question: is there a good general answer to "read a filename from a file and try to use it" for Windows? I found all this surprisingly complex and fiddly.

I guess one option is to not make all this OsString - even though all the low-level arg processing is done with OsString (because the original code used env::args_os), the top-level code immediately tries to utf8ify everything and just bails if that fails. So could we do everything in utf8 from the outset?

Reply:
On Windows there's not really a great option AFAIK, no. On Windows the "best" thing to do is to enforce utf-8 and deal with it that way. We arguably should do the same on Unix to be consistent, but I think it's fine to be specific in this case.